rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,643 @@
+/// !REVIEW: I've tried to closely follow SymCrypts API calls and semantics making some simplifications, but I think we have a chance to simplify further if we choose to.
+/// My questions / Suggestions:
+/// 1. We've introduced the idea of the RsaKeyPair / RsaPublicKey state, which can hold a lot of data, and makes a lot of the calls to the underlying SymCrypt code redundant. most of the accessors no longer
+/// needed to call SymCrypt since the same data is stored on the state or is assumed
+/// 2. to keep this thread safe, I've combined the alloc / generate and alloc / set pattern to just one function
+/// 3. This code is pretty raw, and would not be very usable, the idea would be to write a crate on top of this one, called something like symcrypt_rsa_parse, which will be a friendlier
+/// interface on top of the RSA code we have right now it would have convenience functions like rsa_from_pkcs8, rsa_from_der, rsa_from_asn1 etc. ( this could be part of the intern project as well.)
+/// 4. with how many steps / options there is for generating a key, we can maybe do a builder pattern, this would be similar to how it is on NCrypt / BCrypt where they have
+/// RsaKey functions and the RsaKey.finalize() at the end.
+/// ex:
+///         builder = RsaKeyBuilder::new()
+///             .key_mode(RsaKeyMode::KeyPair)
+///             .key_usage(RsaKeyUsage::Sign)
+///             .generate_key()
+/// My thought is to leave it how it is right now as there are not that many parameters that we need to specify, but thought I would point out the pattern.
+///
+/// 5.
+// !REVIEW: instead of using in/out buffers for the get functions, we can use Vec<> which will be slower but more robust, and will not require the user to know the len of the fields,
+//  Or we could instead return a struct with the info
+//       pub struct KeyReturnValue {
+//          key_mode : RsaKeyMode,
+//          modulus: u64,
+//          p: &[u8],
+//          q: &[u8],
+//              ...
+//        }
+
+///
+// Rsa functions. For further documentation please refer to symcrypt.h
+// TODO rest of the documentation
+use crate::errors::SymCryptError;
+use crate::NumberFormat;
+use std::ptr;
+
+/// 2^16 + 1
+const DEFAULT_PUBLIC_EXPONENT: u64 = 0x10001;
+
+/// Rsa Public and Private Key State.
+///
+/// [`RsaKeyPair`] stores a private and public Rsa key.
+pub struct RsaKeyPair {
+    inner: symcrypt_sys::PSYMCRYPT_RSAKEY,
+    key_usage: RsaKeyUsage,
+}
+
+/// Rsa Public Key State
+///
+/// [`RsaPublicKey`] stores an Rsa public key.
+pub struct RsaPublicKey {
+    inner: symcrypt_sys::PSYMCRYPT_RSAKEY,
+    key_usage: RsaKeyUsage,
+}
+
```

Comment:
These structs are basically the same, just with different names. In SymCrypt we represent RSA keys using one structure, which has a member variable indicating whether or not it has a private key. Is there a particular reason you don't want to do that here?